File: rust/fluentcards/src/enums.rs
use std::fmt;

macro_rules! string_enum {
    (
        $(#[$meta:meta])*
        $name:ident {
            $( $(#[$vmeta:meta])* $variant:ident => $str:literal ),+ $(,)?
        }
    ) => {
        $(#[$meta])*
        #[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
        pub enum $name {
            $( $(#[$vmeta])* $variant ),+
        }

        impl $name {
            /// Returns the string representation used in Adaptive Cards JSON.
            pub fn as_str(&self) -> &'static str {
                match self {
                    $( $name::$variant => $str ),+
                }
            }
        }

        impl AsRef<str> for $name {
            fn as_ref(&self) -> &str {
                self.as_str()
            }
        }

        impl fmt::Display for $name {
            fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
                f.write_str(self.as_str())
            }
        }

        impl From<$name> for serde_json::Value {
            fn from(val: $name) -> serde_json::Value {
                serde_json::Value::String(val.as_str().to_string())
            }
        }
    };
}

string_enum! {
    /// Controls the font size of text in TextBlock and TextRun elements.
    TextSize {
        Small => "small",
        Default => "default",
        Medium => "medium",
        Large => "large",
        ExtraLarge => "extraLarge",
    }
}

string_enum! {
    /// Controls the font weight of text in TextBlock and TextRun elements.
    TextWeight {
        Lighter => "lighter",
        Default => "default",
        Bolder => "bolder",
    }
}

string_enum! {
    /// Controls the color of text in TextBlock and TextRun elements.
    TextColor {
        Default => "default",
        Dark => "dark",
        Light => "light",
        Accent => "accent",
        Good => "good",
        Attention => "attention",
        Warning => "warning",
        White => "white",
    }
}

string_enum! {
    /// Controls the font family used for text rendering.
    FontType {
        Default => "default",
        Monospace => "monospace",
    }
}

string_enum! {
    /// Controls the visual style of a TextBlock.
    TextBlockStyle {
        Default => "default",
        Heading => "heading",
    }
}

string_enum! {
    /// Controls horizontal alignment of elements.
    HorizontalAlignment {
        Left => "left",
        Center => "center",
        Right => "right",
    }
}

string_enum! {
    /// Controls vertical alignment of content within containers.
    VerticalAlignment {
        Top => "top",
        Center => "center",
        Bottom => "bottom",
    }
}

string_enum! {
    /// Controls the spacing before an element.
    Spacing {
        Default => "default",
        None => "none",
        Small => "small",
        Medium => "medium",
        Large => "large",
        ExtraLarge => "extraLarge",
        Padding => "padding",
    }
}

string_enum! {
    /// Controls the visual style of Container and Column elements.
    ContainerStyle {
        Default => "default",
        Emphasis => "emphasis",
        Good => "good",
        Attention => "attention",
        Warning => "warning",
        Accent => "accent",
    }
}

string_enum! {
    /// Controls the display size of Image elements.
    ImageSize {
        Auto => "auto",
        Stretch => "stretch",
        Small => "small",
        Medium => "medium",
        Large => "large",
    }
}

string_enum! {
    /// Controls the shape/style of Image elements.
    ImageStyle {
        Default => "default",
        Person => "person",
    }
}

string_enum! {
    /// Controls the visual style of action buttons.
    ActionStyle {
        Default => "default",
        Positive => "positive",
        Destructive => "destructive",
    }
}

string_enum! {
    /// Controls whether an action appears in the primary or overflow menu.
    ActionMode {
        Primary => "primary",
        Secondary => "secondary",
    }
}

string_enum! {
    /// Controls the keyboard type shown for Input.Text on mobile devices.
    TextInputStyle {
        Text => "text",
        Tel => "tel",
        Url => "url",
        Email => "email",
        Password => "password",
    }
}

string_enum! {
    /// Controls the display style of Input.ChoiceSet.
    ChoiceInputStyle {
        Compact => "compact",
        Expanded => "expanded",
        Filtered => "filtered",
    }
}

string_enum! {
    /// Controls where the label is rendered relative to the input.
    InputLabelPosition {
        Inline => "inline",
        Above => "above",
    }
}

string_enum! {
    /// Controls how password-style inputs reveal their content.
    InputStyle {
        Default => "default",
        RevealOnHover => "revealOnHover",
    }
}

string_enum! {
    /// Controls which inputs are submitted with an action.
    AssociatedInputs {
        Auto => "auto",
        None => "none",
    }
}

string_enum! {
    /// Controls how a background image fills its container.
    BackgroundImageFillMode {
        Cover => "cover",
        RepeatHorizontally => "repeatHorizontally",
        RepeatVertically => "repeatVertically",
        Repeat => "repeat",
    }
}

string_enum! {
    /// Indicates the severity level of a validation issue.
    ValidationSeverity {
        Info => "info",
        Warning => "warning",
        Error => "error",
    }
}

File: swift/Sources/FluentCards/Enums.swift
/// TextSize controls the font size of text in TextBlock and TextRun elements.
public enum TextSize: String {
    case small = "small"
    case `default` = "default"
    case medium = "medium"
    case large = "large"
    case extraLarge = "extraLarge"
}

/// TextWeight controls the font weight of text in TextBlock and TextRun elements.
public enum TextWeight: String {
    case lighter = "lighter"
    case `default` = "default"
    case bolder = "bolder"
}

/// TextColor controls the color of text in TextBlock and TextRun elements.
public enum TextColor: String {
    case `default` = "default"
    case dark = "dark"
    case light = "light"
    case accent = "accent"
    case good = "good"
    case attention = "attention"
    case warning = "warning"
    case white = "white"
}

/// FontType controls the font family used for text rendering.
public enum FontType: String {
    case `default` = "default"
    case monospace = "monospace"
}

/// TextBlockStyle controls the visual style of a TextBlock.
public enum TextBlockStyle: String {
    case `default` = "default"
    case heading = "heading"
}

/// HorizontalAlignment controls horizontal alignment of elements.
public enum HorizontalAlignment: String {
    case left = "left"
    case center = "center"
    case right = "right"
}

/// VerticalAlignment controls vertical alignment of content within containers.
public enum VerticalAlignment: String {
    case top = "top"
    case center = "center"
    case bottom = "bottom"
}

/// Spacing controls the spacing before an element.
public enum Spacing: String {
    case `default` = "default"
    case none = "none"
    case small = "small"
    case medium = "medium"
    case large = "large"
    case extraLarge = "extraLarge"
    case padding = "padding"
}

/// ContainerStyle controls the visual style of Container and Column elements.
public enum ContainerStyle: String {
    case `default` = "default"
    case emphasis = "emphasis"
    case good = "good"
    case attention = "attention"
    case warning = "warning"
    case accent = "accent"
}

/// ImageSize controls the display size of Image elements.
public enum ImageSize: String {
    case auto = "auto"
    case stretch = "stretch"
    case small = "small"
    case medium = "medium"
    case large = "large"
}

/// ImageStyle controls the shape/style of Image elements.
public enum ImageStyle: String {
    case `default` = "default"
    case person = "person"
}

/// ActionStyle controls the visual style of action buttons.
public enum ActionStyle: String {
    case `default` = "default"
    case positive = "positive"
    case destructive = "destructive"
}

/// ActionMode controls whether an action appears in the primary or overflow menu.
public enum ActionMode: String {
    case primary = "primary"
    case secondary = "secondary"
}

/// TextInputStyle controls the keyboard type shown for Input.Text on mobile devices.
public enum TextInputStyle: String {
    case text = "text"
    case tel = "tel"
    case url = "url"
    case email = "email"
    case password = "password"
}

/// ChoiceInputStyle controls the display style of Input.ChoiceSet.
public enum ChoiceInputStyle: String {
    case compact = "compact"
    case expanded = "expanded"
    case filtered = "filtered"
}

/// InputLabelPosition controls where the label is rendered relative to the input.
public enum InputLabelPosition: String {
    case inline = "inline"
    case above = "above"
}

/// InputStyle controls how password-style inputs reveal their content.
public enum InputStyle: String {
    case `default` = "default"
    case revealOnHover = "revealOnHover"
}

/// AssociatedInputs controls which inputs are submitted with an action.
public enum AssociatedInputs: String {
    case auto = "auto"
    case none = "none"
}

/// BackgroundImageFillMode controls how a background image fills its container.
public enum BackgroundImageFillMode: String {
    case cover = "cover"
    case repeatHorizontally = "repeatHorizontally"
    case repeatVertically = "repeatVertically"
    case `repeat` = "repeat"
}

/// ValidationSeverity indicates the severity level of a validation issue.
public enum ValidationSeverity: String {
    case info = "info"
    case warning = "warning"
    case error = "error"
}
